### PR TITLE
CSS Naming Convention

### PR DESCRIPTION
**Pull request by simurai**
# CSS Naming Convention

**Question:** Should Atom's CSS use a naming convention?

👉  [Rendered](https://github.com/atom/design-decisions/blob/css-naming-convention/css-naming-convention.md)

This is part of the lager [Theming in Atom](https://github.com/atom/design-decisions/pull/4) proposal.

## Decision

TBD

### CODE REVIEW

**Comment by Alhadis** [comment]
You guys should be abusing Less to scope keybindings and editor configs:

So instead of [`base.cson`](https://github.com/atom/atom/blob/c100ad0108e63049c879e4bcb6877f85cae28f42/keymaps/base.cson#L1), you'd have this `base.css`:

~~~less
/* Platform Bindings */
atom-text-editor {
	home: "editor move-to-first-character-of-line";
	end:  "editor move-to-end-of-screen-line";
	shift-home: "editor select-to-first-character-of-line";
	shift-end:  "editor select-to-end-of-line";

	&:not([mini]) {
		ctrl-shift-c: "editor copy-path";
		tab: "editor indent";
		enter: "editor newline";
		shift-tab: "editor outdent-selected-rows";
		ctrl-shift-k: "editor delete-line";
	}

	.select-list &[mini] {
		enter: "core confirm";
	}
}

.tool-panel.panel-left,
.tool-panel.panel-right {
	escape: "tool-panel unfocus";
}

/* Duplicate selectors won't overwrite each other */
atom-text-editor {
	
	/* Special keywords don't need quotes: */
	enter: "native!";  // Bad (Legacy CSON way)
	enter: native;     // Good
	enter: initial;    // Best

	/* Proper way to use !important: */
	enter: initial !important;
}
~~~

It's weird as hell, but I'll be damned if that isn't the cleanest way to express scope-based settings. Particularly if the future alternative is JSON, because blegh, screw key-quoting).

**Comment by simurai** [comment]
> It's weird as hell

Hehe.. yeah. It might also be harder to learn for people that are not familiar with Less.

A thought: Should keybindings get their unique identifiers? Because even if we find a compact BEM name for that element, like `.select-list atom-text-editor[mini]` -> `SelectList-input`, we would be stuck again for the next few years and can't change that class name without breaking keybindings.

So instead, we would use a class or attribute **only** for keybindings and context menu scoping. For example 

```html
<atom-text-editor key="select-list-input">
```
```
'[key="select-list-input"]':
  'enter': 'core:confirm'
```

Package and theme authors would never use these `key` attributes for styling and they are reserved only for keybindings + context menu. A bit more typing when authoring the markup, but then they can change individually without breaking each other.

File: css-naming-convention.md [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+This is Atom's naming convention for creating UI in Atom and Atom packages. It's close to [BEM/SUIT](https://github.com/suitcss/suit/blob/master/doc/naming-conventions.md), but slightly customized for Atom's use case.
+
+
+## Example
+
+Below the commit box as a possible example:
+
+```html
+<div class='gitub-CommitBox'>
+	<div class='gitub-CommitBox-editor'></div>
```
Comment (Arcanemagus) [inline css-naming-convention.md]: Inconsistent indentation here, this whole inner block needs to lose 6 spaces 😉.

File: css-naming-convention.md [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+.github-CommitBox-commitButton.is-disabled {
+  // yep
+}
+```
+
+> Note: If multiple words are needed, camelCase is used: `has-collapsedItems`.
+
+
+
+## More guidelines, not crucial, but still good to know and follow. ;-)
```
Comment (Arcanemagus) [inline css-naming-convention.md]: I'd probably shorten this to just "More guidelines" if this is going to be a formal document 😉.

File: css-naming-convention.md [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+- Styling elements (like `div`) should be avoided. This makes it easier to switch elements, like from a `<button>` to `<a>`.
+- No utility classes. Themes and user styles can only override CSS but not change the markup. Therefore having utility classes doesn't make as much sense once you override them.
+- Avoid using existing CSS classes to reference elements in the DOM. That way when the markup changes, functionality and specs will less likely break. Instead use `ref` attributes (`ref="commitButton"`).
+- Avoid changing modifiers at runtime. If you're in need, consider changing the modifier class into a state. For example if you often want to switch a default button to a primary button, change the `github-CommitBox-commitButton--primary` modifier class into a `is-primary` state instead. Since state classes are decoupled from components, it's easier to change component names, but keep the states the same.
```
Comment (Arcanemagus) [inline css-naming-convention.md]: I _think_ "component names" and "states" got switched in the last sentence here? I'm mostly just guessing here though as I'm very _un_ familiar with CSS.

File: css-naming-convention.md (hunk 1 above)
Comment (Alhadis) [reply css-naming-convention.md]: It's a tab. It shows up as 8 spaces to you because you haven't [installed this](https://chrome.google.com/webstore/detail/tab-size-on-github/ofjbgncegkdemndciafljngjbdpfmbkn?utm_source=chrome-app-launcher-info-dialog), probably. :stuck_out_tongue_closed_eyes:

Comment (Arcanemagus) [reply css-naming-convention.md]: You're missing the bigger picture that whatever it is, it's not consistent through the entire block.

File: css-naming-convention.md (hunk 3 above)
Comment (simurai) [reply css-naming-convention.md]: Oh, I see. It's actually easier to change both.

- Change component name: `github-CommitBox-commitButton is-disabled` -> `github-Button is-disabled`
- Change state name: `github-CommitBox-commitButton is-disabled` -> `github-CommitBox-commitButton is-inActive`

I'll rewrite it to mention both. Thanks. 👍

**Comment by simurai** [comment]
Closing this again.. we started to use a similar naming convention for the GitHub package https://github.com/atom/github/pull/1379. But for Atom it would only make sense in a major release, where it's ok to break themes/packages.
